app/src/main/java/com/github/sdpteam15/polyevents/view/fragments/LeaveEventReviewFragment.kt

```diff
@@ -0,0 +1,157 @@
+
+    private var rated = false
+
+    override fun onCreateView(
+        inflater: LayoutInflater,
+        container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View? {
+        val view = super.onCreateView(inflater, container, savedInstanceState)
+        if (view != null) {
+            leaveReviewDialogConfirmButton = view.findViewById(R.id.leave_review_fragment_save_button)
+
+            userFeedbackDialogEditText = view.findViewById(R.id.leave_review_fragment_feedback_text)
+
+            leaveReviewDialogRatingBar = view.findViewById(R.id.leave_review_fragment_rating)
+            leaveReviewDialogRatingBar.setOnRatingBarChangeListener { _, _, _ ->
+                rated = true
+            }
+
+            leaveReviewDialogCancelButton = view.findViewById(R.id.leave_review_fragment_cancel_button)
+            leaveReviewDialogCancelButton.setOnClickListener {
+                // Dimiss the dialog if canceled
+                dismiss()
+            }
+
+            leaveReviewDialogConfirmButton.setOnClickListener {
+                onClickAdd()
+            }
+
+            val ratingObservable = Observable<Rating>()
+            ratingObservable.observe(this) {
+                Log.d(TAG, "Retrieved Rating for $eventId and ${currentDatabase.currentUser!!.uid}!")
+                val rating = it.value
+                leaveReviewDialogRatingBar.rating = rating.rate!!
+                if (rating.feedback != null) {
+                    if (!rating.feedback.isEmpty()) {
+                        userFeedbackDialogEditText.setText(rating.feedback)
+                    }
+                }
+                leaveReviewDialogConfirmButton.setOnClickListener {
+                    onClickUpdate(rating)
+                }
+            }
+
+            currentDatabase.eventDatabase!!.getUserRatingFromEvent(
+                userId = currentDatabase.currentUser!!.uid,
+                eventId = eventId!!,
+                returnedRating = ratingObservable,
+                userAccess = null
+            )
+        }
+
+        return view
+    }
+
+    override fun onResume() {
+        super.onResume()
+        // Uncomment the following lines to manually set the height and the width of the DialogFragment
+        /*val width = ConstraintLayout.LayoutParams.MATCH_PARENT
+        val height = ConstraintLayout.LayoutParams.MATCH_PARENT
+        dialog!!.window!!.setLayout(width, height)*/
+    }
+
+    /**
+     * A listener to update a rating in the database. Used when we have an already found
+     * rating for the current user.
+     */
+    private fun onClickUpdate(rating: Rating) {
+        if (!rated) {
+            // Check if user has rated, to avoid storing a rating with zero stars.
+            HelperFunctions.showToast(getString(R.string.event_review_leave_rating_warning),
+                context
+            )
+        } else {
+            Log.d(TAG, "Leaving a review")
+            val ratingCopy = rating.copy(
```

Comment:
Nice use of copy, didn't know about this function :)

Reply:
Yes! Very useful and implemented by default if your class is declared a data class.